File: tests/test_slide.py
"""The graph slide clock is JS; this runs its self-check when node is present."""

import shutil
import subprocess
from pathlib import Path

import pytest

REPO = Path(__file__).resolve().parent.parent
CHECK = REPO / "tests" / "test_slide.mjs"


@pytest.mark.skipif(shutil.which("node") is None, reason="node is not installed")
def test_slide_clock() -> None:
    completed = subprocess.run(
        ["node", str(CHECK)],
        cwd=REPO,
        check=False,
        capture_output=True,
        text=True,
    )
    assert completed.returncode == 0, completed.stdout + completed.stderr

File: tests/test_slide.mjs
import assert from 'node:assert/strict';
import {
  phaseThroughBucket,
  sizeHeld,
  slideStepRatio,
  timelineAdvanced,
} from '../web/slide.js';

assert.equal(phaseThroughBucket(0, 100, true), 0);
assert.equal(phaseThroughBucket(50, 100, true), 0.5);
assert.equal(phaseThroughBucket(100, 100, false), 1);
assert.equal(phaseThroughBucket(150, 100, false), 1);
assert.equal(phaseThroughBucket(150, 100, true), 0.5);
assert.equal(phaseThroughBucket(0, 0, true), 0);

assert.equal(timelineAdvanced(null, 1000), true);
assert.equal(timelineAdvanced(1000, 1000), false);
assert.equal(timelineAdvanced(1000, 1100), true);
assert.equal(timelineAdvanced(null, null), false);

assert.equal(sizeHeld(100, 100, 1), 100);
assert.equal(sizeHeld(100, 101, 1), 100);
assert.equal(sizeHeld(100, 102, 1), 102);
assert.equal(sizeHeld(0, 50, 1), 50);

assert.equal(slideStepRatio(100), 0.01);
assert.equal(slideStepRatio(0), 0);

console.log('ok');
